File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/js/JsProcessWebApi.kt
package info.bagen.rust.plaoc.microService.sys.js

import android.net.Uri
import android.webkit.WebMessage
import info.bagen.rust.plaoc.microService.helper.Mmid
import info.bagen.rust.plaoc.microService.helper.gson
import info.bagen.rust.plaoc.microService.ipc.IPC_ROLE
import info.bagen.rust.plaoc.microService.ipc.Ipc
import info.bagen.rust.plaoc.microService.ipc.ipcWeb.MessagePortIpc
import info.bagen.rust.plaoc.microService.ipc.ipcWeb.saveNative2JsIpcPort
import info.bagen.rust.plaoc.microService.webview.DWebView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class JsProcessWebApi(val dWebView: DWebView) {
    suspend fun isReady() =
        dWebView.evaluateSyncJavascriptCode("typeof createProcess") == "function"

    data class ProcessInfo(val process_id: Int) {}
    inner class ProcessHandler(val info: ProcessInfo, var ipc: MessagePortIpc)

    suspend fun createProcess(
        env_script_url: String,
        metadata_json: String,
        env_json: String,
        remoteModule: Ipc.MicroModuleInfo,
        host: String
    ) = withContext(Dispatchers.Main) {
        val channel = dWebView.createWebMessageChannel()
        val port1 = channel[0]
        val port2 = channel[1]
        val metadata_json_str = gson.toJson(metadata_json)
        val env_json_str = gson.toJson(env_json)

        val processInfo_json = dWebView.evaluateAsyncJavascriptCode("""
            new Promise((resolve,reject)=>{
                addEventListener("message", async event => {
                    if (event.data === "js-process/create-process") {
                        const fetch_port = event.ports[0];
                        try{
                            resolve(await createProcess(`$env_script_url`,$metadata_json_str,$env_json_str,fetch_port,`$host`))
                        }catch(err){
                            reject(err)
                        }
                    }
                }, { once: true })
            })
            """.trimIndent(), afterEval = {
            try {

                dWebView.postWebMessage(
                    WebMessage("js-process/create-process", arrayOf(port1)), Uri.EMPTY
                );
            } catch (e: Exception) {
                println("QQQQ");
                e.printStackTrace()
            }
        })
        debugJsProcess("processInfo", processInfo_json)
        val info = gson.fromJson(processInfo_json, ProcessInfo::class.java)
        ProcessHandler(info, MessagePortIpc(port2, remoteModule, IPC_ROLE.CLIENT))
    }


    data class RunProcessMainOptions(val main_url: String)

    suspend fun runProcessMain(process_id: Int, options: RunProcessMainOptions) =
        dWebView.evaluateAsyncJavascriptCode(
            """
        runProcessMain($process_id, { main_url:`${options.main_url}` })
        """.trimIndent()
        ).let {}

    suspend fun createIpc(process_id: Int, mmid: Mmid) = withContext(Dispatchers.Main) {
        val channel = dWebView.createWebMessageChannel()
        val port1 = channel[0]
        val port2 = channel[1]
        dWebView.evaluateAsyncJavascriptCode("""
        new Promise((resolve,reject)=>{
            addEventListener("message", async event => {
                if (event.data === "js-process/create-ipc") {
                    const ipc_port = event.ports[0];
                    try{
                        resolve(await createIpc($process_id, `$mmid`, ipc_port))
                    }catch(err){
                        reject(err)
                    }
                }
            }, { once: true })
        })
        """.trimIndent(), afterEval = {
            dWebView.postWebMessage(
                WebMessage("js-process/create-ipc", arrayOf(port1)), Uri.EMPTY
            );
        })
        saveNative2JsIpcPort(port2)
    }

    fun destroy() {
        dWebView.destroy()
    }
}

File: ios/BFExplorer/Plaoc/bfs/microModule/sys/js/JsProcessWebApi.swift
//
//  JsProcessWebApi.swift
//  BFExplorer
//
//  Created by ui08 on 2023/2/27.
//

import Foundation
import WebKit
import Combine

class JsProcessWebApi: NSObject {
    let webView: DWebView
    private let processIdSubject = PassthroughSubject<Int, Never>()
    private var createProcessPort1: MessagePortIpc?
    private var createIpcPort1: WebMessagePort?
    private var cancellable: AnyCancellable?
    private var processIdMmidMap: [Int:String] = [:]
    
    init(webView: DWebView) {
        self.webView = webView
        super.init()
    }
    
    struct ProcessInfo: Codable {
        let process_id: Int
    }
    
    class ProcessHandler {
        let info: ProcessInfo
        var ipc: MessagePortIpc
        
        init(info: ProcessInfo, ipc: MessagePortIpc) {
            self.info = info
            self.ipc = ipc
        }
    }
    
    func createProcess(
        env_script_url: String,
        metadata_json: String,
        env_json: String,
        remoteModule: MicroModuleInfo,
        host: String
    ) async -> ProcessHandler {
        // 创建namespace
        await self.webView.configuration.userContentController.add(LeadScriptHandle(messageHandle: self),
                                                             contentWorld: .world(name: remoteModule.mmid),
                                                             name: "onmessage")
        await self.webView.configuration.userContentController.add(LeadScriptHandle(messageHandle: self),
                                                             contentWorld: .world(name: remoteModule.mmid),
                                                             name: "processId")
        await self.webView.configuration.userContentController.add(LeadScriptHandle(messageHandle: self),
                                                             contentWorld: .world(name: remoteModule.mmid),
                                                             name: "logging")
        
        
        _ = await self.webView.callAsyncJavaScript("""
            const {port1, port2} = new MessageChannel();
            port2.onmessage = (evt) => {
                window.webkit.messageHandlers.onmessage.postMessage(evt.data);
            }
            await new Promise((resolve)=>{self.createProcess_start = resolve});
            try {
                let { process_id } = await createProcess(env_script_url, metadata_json_str, env_json_str, port1, host);
                window.webkit.messageHandlers.processId.postMessage(`${process_id}`);
            } catch(err) {
                window.webkit.messageHandlers.logging.postMessage(err.message ?? JSON.stringify(err));
            }
        """.trimmingCharacters(in: .whitespacesAndNewlines), arguments: [
            "env_script_url": env_script_url,
            "metadata_json_str": metadata_json,
            "env_json_str": env_json,
            "host": host
        ], in: nil, in: .world(name: remoteModule.mmid))
        
        return await withCheckedContinuation { continuation in
            cancellable = processIdSubject.sink(receiveValue: { process_id in
                self.processIdMmidMap[process_id] = remoteModule.mmid
                self.createProcessPort1 = MessagePortIpc(port: .init(name: "\(process_id)", role: .port1), remote: remoteModule, role_type: .server)
                
                let processHandler = ProcessHandler(
                    info: .init(process_id: process_id),
                    ipc: .init(port: .init(name: "\(process_id)", role: .port2), remote: remoteModule, role_type: .client))

                continuation.resume(returning: processHandler)
            })
            
        }
    }
    
    struct RunProcessMainOptions {
        let main_url: String
    }
    
    func runProcessMain(process_id: Int, options: RunProcessMainOptions) async {
        let mmid = processIdMmidMap[process_id]
        await self.webView.callAsyncJavaScript(
            "runProcessMain(process_id, { main_url: main_url });",
            arguments: ["process_id": process_id, "main_url": options.main_url],
            in: nil,
            in: .world(name: mmid!))
    }
    
    func createIpc(process_id: Int, mmid: Mmid) async -> Int {
        self.processIdMmidMap[process_id] = mmid
        self.createIpcPort1 = WebMessagePort(name: "createIpc_\(process_id)", role: .port1)
        let port2 = WebMessagePort(name: "createIpc_\(process_id)", role: .port2)
        await self.webView.callAsyncJavaScript("""
            const {port1_ipc, port2_ipc} = new MessageChannel();
            port2_ipc.onmessage = (evt) => {
                window.webkit.messageHandlers.ipcOnMessage.postMessage(JSON.stringify(evt.data));
            }
            await createIpc(process_id, mmid, port1_ipc);
        """, arguments: ["process_id": process_id, "mmid":mmid], in: nil, in: .world(name: mmid))
        
        return saveNative2JsIpcPort(port: port2)
    }
    
    deinit {
        processIdMmidMap.removeAll()
        cancellable?.cancel()
    }
}

extension JsProcessWebApi: WKScriptMessageHandler {
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "onmessage" {
            print("onmessage")
            if createProcessPort1 != nil {
                guard let body = message.body as? String else { return }
                let ipcMessage = jsonToIpcMessage(data: body, ipc: createProcessPort1!)
                if ipcMessage != nil {
                    Task {
                        await createProcessPort1!.postMessage(message: ipcMessage!)
                    }
                }
            }
        } else if message.name == "processId" {
            print("processId")
            guard let process_id = message.body as? Int else  {
                return
            }
            
            processIdSubject.send(process_id)
        } else if message.name == "logging" {
            print("logging")
            print(message.body)
        } else if message.name == "ipcOnMessage" {
            print("ipcOnMessage")
            if createIpcPort1 != nil {
                guard let body = message.body as? String else { return }
                createIpcPort1!.postMessage(body)
            }
        }
    }
}
